populations/models.py

```diff
@@ -149,52 +105,93 @@ def find_submodels(name, obj):
     # hyperparameters in any model
     Nhyper = np.max([len(x.split('/')) for x in hyperparams])
 
-    # construct hyperparam dict, with keys [0,..,Nhyper] and
-    # values the points in hyperparameter space with simulations
     while hyperidx < Nhyper:
         hyperidx_with_Nhyper = np.argwhere(np.asarray([len(x.split('/')) \
                 for x in hyperparams])>hyperidx).flatten()
         hyperparams_at_level = sorted(set([x.split('/')[hyperidx] \
                 for x in np.asarray(hyperparams)[hyperidx_with_Nhyper]]))
         hyperparam_dict[hyperidx] = hyperparams_at_level
         hyperidx += 1
+    # length of the hyperparam dict for each dimension
+    hyperparam_pts_per_dim = [len(hyperparam_dict[x]) for x in range(Nhyper)]
+    return deepest_models, hyperparam_pts_per_dim
+
+def get_channel_smdls(chnl, deepest_models, hyperparam_pts_per_dim):
+
+    # find submodel keys, and indices they should correspond to
+    #   in the input samples dict to the FlowModel
+    channel_smdls = [x for x in deepest_models if chnl+'/' in x]
+    channel_smdls_split = np.array([x.split('/')[1:] \
+                for x in deepest_models if chnl+'/' in x])
+    smdl_indices = [list(np.arange(hyperparam_pts_per_dim[i])) \
+                for i in range(channel_smdls_split.shape[1])]
+    smdl_indxs_combos = np.squeeze(list(product(*smdl_indices)))
+
+    return channel_smdls, smdl_indxs_combos
 
-    # KDE case: reads in submodel for each of the deepest
-    #   model and sends to KDEModel
-    # Flow case: reads in samples from all channels
-    #   and sends to FlowModel
+def get_popsynth_outputs(file_path, chnl, channel_dict, param_dict, spinmag):
+    deepest_models, hyperparam_pts_per_dim = get_deepest_models(file_path, channel_dict)
+    channel_smdls, smdl_indxs_combos = get_channel_smdls(chnl, deepest_models, hyperparam_pts_per_dim)
+    popsynth_outputs = read_hdf5(file_path, chnl, channel_smdls, smdl_indxs_combos, param_dict, spinmag)
+    return popsynth_outputs
+
+def get_models(file_path, channel_dict, param_dict, \
+            hyperparam_dict, use_flows, \
+            sensitivity=None, **kwargs):
+    """
+    Call this to get all the models and submodels, as well
+    as KDEs of these models, packed inside of dictionaries labelled in the
+    dict structure models[channel][smdl]. Will first look for :params: as
+    series in the dataframe. If they are not present, it will try to construct
+    these parameters if the valid transformations are present in transforms.py.
+
+    Parameters
+    ----------
+    file_path : str
+        filepath to models_reduced.hdf5
+    channel_dict : dict with channel names as keys
+        values contain 'parameters' and 'fullname'
+    param_dict : dict with event-level parameters as keys, and limits and
+        full names as values
+    hyperparam_dict : dict with population hyperparameters as keys, and
+        discrete values (with value key)/full names as values
+    use_flows : bool
```

Comment:
Can you add what sensitivity is as well as what kwargs are available? I want to make sure that it can still read in spinmag, max_samps, and kde_bandwidth. The latter two should have defaults if they aren't specified in kwargs.